### ISSUE TITLE
how to include loguru?

### ISSUE THREAD

**Issue by mcauto** [marcelotduarte/cx_Freeze#745]
🥺

**setup.py**
```python
build_exe_options = {
    "zip_include_packages": ["*"],
    "zip_exclude_packages": [],
    "include_files": ["test-infra/sftp", "log"],
    "includes": [
        "_cffi_backend",
        "encodings.idna",
        "uvloop._noop",
        "loguru",
        "sysconfig",
    ],
}
```

**output**

```bash
Traceback (most recent call last):
  File "/root/projects/py/ims/.venv/lib/python3.7/site-packages/cx_Freeze/initscripts/__startup__.py", line 40, in run
    module.run()
  File "/root/projects/py/ims/.venv/lib/python3.7/site-packages/cx_Freeze/initscripts/Console.py", line 33, in run
    exec(code,  m.__dict__)
  File "gabia_imsd/__main__.py", line 4, in <module>
  File "/root/projects/py/ims/.venv/lib/python3.7/site-packages/loguru/__init__.py", line 20, in <module>
    logger.add(_sys.stderr)
  File "/root/projects/py/ims/.venv/lib/python3.7/site-packages/loguru/_logger.py", line 936, in add
    prefix=exception_prefix,
  File "/root/projects/py/ims/.venv/lib/python3.7/site-packages/loguru/_better_exceptions.py", line 133, in __init__
    self._lib_dirs = self._get_lib_dirs()
  File "/root/projects/py/ims/.venv/lib/python3.7/site-packages/loguru/_better_exceptions.py", line 142, in _get_lib_dirs
    paths = {sysconfig.get_path(name, scheme) for scheme in schemes for name in names}
  File "/root/projects/py/ims/.venv/lib/python3.7/site-packages/loguru/_better_exceptions.py", line 142, in <setcomp>
    paths = {sysconfig.get_path(name, scheme) for scheme in schemes for name in names}
  File "/usr/lib/python3.7/sysconfig.py", line 513, in get_path
    return get_paths(scheme, vars, expand)[name]
  File "/usr/lib/python3.7/sysconfig.py", line 503, in get_paths
    return _expand_vars(scheme, vars)
  File "/usr/lib/python3.7/sysconfig.py", line 172, in _expand_vars
    _extend_dict(vars, get_config_vars())
  File "/usr/lib/python3.7/sysconfig.py", line 551, in get_config_vars
    _init_posix(_CONFIG_VARS)
  File "/usr/lib/python3.7/sysconfig.py", line 422, in _init_posix
    _temp = __import__(name, globals(), locals(), ['build_time_vars'], 0)
ModuleNotFoundError: No module named '_sysconfigdata_m_linux_x86_64-linux-gnu'
```

**Comment by marcelotduarte** [comment]
If ou hava a C compiler, you can use the current master.
`pip install -U git+https://github.com/marcelotduarte/cx_Freeze.git@master`

**Comment by mcauto** [comment]
@marcelotduarte Thank you! save time!

I want to use [pydantic](https://github.com/samuelcolvin/pydantic)

**setup.py**

```python
build_exe_options = {
    "zip_include_packages": ["*"],
    "zip_exclude_packages": [],
    "include_files": ["test-infra/sftp", "log"],
    "includes": [
        "_cffi_backend",
        "encodings.idna",
        "uvloop._noop",
        "loguru",
        "pydantic",
    ],
}
```
But It's not working.. help me 😳

```bash
...
from pydantic import BaseSettings, Field
ModuleNotFoundError: No module named 'pydantic'
```

**Comment by marcelotduarte** [comment]
Use packages (See at https://cx-freeze.readthedocs.io/en/latest/distutils.html#build-exe)
In this case, "pydantic" is a package, "pydantic.fields" and "pydantic.errors" are modules.

In your options, I see that "encodings.idna", "uvloop._noop", are detected by the current version.

**Comment by mcauto** [comment]
@marcelotduarte 

Thank you for the kind explanation 👍

**Comment by marcelotduarte** [comment]
You're welcome!

**Comment by mcauto** [comment]
@marcelotduarte 

it's not working.. 😨

**setup.py**
```python
build_exe_options = {
    "zip_include_packages": ["*"],
    "zip_exclude_packages": [],
    "include_files": ["test-infra/sftp", "log"],
    "includes": ["_cffi_backend", "encodings.idna", "uvloop._noop", "loguru"],
    "packages": ["pydantic"],
}
```

**building_exe**

```bash
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/__init__.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/__init__.cpython-37m-x86_64-linux-gnu.so
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/class_validators.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/pydantic.class_validators.cpython-37m-x86_64-linux-gnu.so
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/color.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/pydantic.color.cpython-37m-x86_64-linux-gnu.so
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/dataclasses.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/pydantic.dataclasses.cpython-37m-x86_64-linux-gnu.so
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/datetime_parse.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/pydantic.datetime_parse.cpython-37m-x86_64-linux-gnu.so
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/decorator.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/pydantic.decorator.cpython-37m-x86_64-linux-gnu.so
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/env_settings.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/pydantic.env_settings.cpython-37m-x86_64-linux-gnu.so
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/error_wrappers.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/pydantic.error_wrappers.cpython-37m-x86_64-linux-gnu.so
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/errors.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/pydantic.errors.cpython-37m-x86_64-linux-gnu.so
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/fields.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/pydantic.fields.cpython-37m-x86_64-linux-gnu.so
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/json.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/pydantic.json.cpython-37m-x86_64-linux-gnu.so
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/main.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/pydantic.main.cpython-37m-x86_64-linux-gnu.so
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/mypy.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/pydantic.mypy.cpython-37m-x86_64-linux-gnu.so
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/networks.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/pydantic.networks.cpython-37m-x86_64-linux-gnu.so
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/parse.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/pydantic.parse.cpython-37m-x86_64-linux-gnu.so
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/schema.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/pydantic.schema.cpython-37m-x86_64-linux-gnu.so
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/tools.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/pydantic.tools.cpython-37m-x86_64-linux-gnu.so
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/types.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/pydantic.types.cpython-37m-x86_64-linux-gnu.so
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/typing.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/pydantic.typing.cpython-37m-x86_64-linux-gnu.so
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/utils.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/pydantic.utils.cpython-37m-x86_64-linux-gnu.so
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/validators.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/pydantic.validators.cpython-37m-x86_64-linux-gnu.so
copying /root/projects/py/ims/.venv/lib/python3.7/site-packages/pydantic/version.cpython-37m-x86_64-linux-gnu.so -> build/exe.linux-x86_64-3.7/lib/pydantic.version.cpython-37m-x86_64-linux-gnu.so
```

```bash
./my_exe
...
    from pydantic.env_settings import BaseSettings
ModuleNotFoundError: No module named 'pydantic'
```

**Comment by marcelotduarte** [comment]
I did a test with py 3.8 on Linux. I used the sample at https://github.com/samuelcolvin/pydantic#a-simple-example
I used the command to get success: 
`cxfreeze test_pydantic.py -c --includes=json,ipaddress,uuid,colorsys`
The reason for cx-freeze does not discover the packages/modules used by pydantic is that it is compiled by cython.
You have two options in this case:
1) Install pydantic from sources and cx-freeze will discover the dependencies
2) Use the fast compiled version and manually include the dependencies of modules in includes/packages option.

My result:
(pd) marcelo@teste6:~/D/testes/pd$ dist/test_pydantic 
id=123 signup_ts=datetime.datetime(2017, 6, 1, 12, 22) friends=[1, 2, 3] name='John Doe'
123

**Comment by mcauto** [comment]
@marcelotduarte 

Thank you for your time!! 😀

> 1. Install pydantic from sources and cx-freeze will discover the dependencies

I don't understand.. 😳 how to install pydantic from sources?

> 2. Use the fast compiled version and manually include the dependencies of modules in includes/packages option.

Is it right for your explanation? 

**setup.py**

```python
pydantic_files: List[str] = get_python_files(
    root_path=f"{os.getcwd()}/.venv/lib/python3.7/site-packages/pydantic/"
) # ['utils.py', 'datetime_parse.py', 'generics.py', 'mypy.py', 'main.py', 'errors.py', 'typing.py', 'dataclasses.py', 'networks.py', 'validators.py', 'version.py', 'color.py', 'schema.py', 'json.py', 'class_validators.py', 'parse.py', 'tools.py', 'types.py', 'fields.py', 'decorator.py', '__init__.py', 'error_wrappers.py', 'env_settings.py']
...
setup(
    ...
# add includes/packages
```

**Comment by marcelotduarte** [comment]
It is off-topic, but...
1. pip install -U git+https://github.com/samuelcolvin/pydantic.git@v1.6.1
2. No, include the dependent modules, not the modules of pydantic. pydantic depend of json,ipaddress,uuid,colorsys and probably others... then:
build_exe_options = {
    "zip_include_packages": ["*"],
    "zip_exclude_packages": [],
    "include_files": ["test-infra/sftp", "log"],
    "includes": ["_cffi_backend", ""loguru", "json", "ipaddress", "uuid", "colorsys"],

}

**Comment by mcauto** [comment]
@marcelotduarte  
It's works! 
Thank you! so kindful 🥺 Have a nice day!

**Comment by marcelotduarte** [comment]
cx_Freeze 6.7b2 has just been released with hooks to support pydantic (#1074).
pip install cx-Freeze==6.7b2
